### ISSUE TITLE
Span<T> methods may not inline when called from shared methods

### ISSUE THREAD

**Issue by AndyAyersMS** [dotnet/coreclr#10031]
Several of `Span<T>`'s methods invoke methods on other generic classes, eg `ByReference<T>` or `Unsafe<T>`. If the span methods are in turn invoked from a method where `T` requires runtime lookup, this need for runtime lookup passes transitively to generic callees made by the span method. Unfortunately this transitive need blocks inlining of the span methods as CoreCLR is not able to support inlining methods that perform runtime lookups. Inlining is blocked even when the span methods are given the `AggressiveInlining` attribute.

We can get around this by implementing (where possible) the generic methods invoked by Span as jit intrinsics. For example the span indexer calls `Unsafe.Add<T>(ref T, int)`  and that method only depends on `sizeof(T)` which is is known to be 4 or 8 for shared methods (since sharing implies `T` is a ref type and so is pointer sized).

**Comment by RussKeldorph** [comment]
@russellhadley @AndyAyersMS Can we get this assigned to someone or moved out of 2.0?

**Comment by russellhadley** [comment]
Assigning you @AndyAyersMS.  You can triage it further.

**Comment by AndyAyersMS** [comment]
Simple example: we'd like the span ctor and indexer to inline into `X<T>.C`, when `T` is a ref type.
```C#
using System;
using System.Runtime.CompilerServices;

class X<T>
{
    [MethodImpl(MethodImplOptions.NoInlining)]
    public static T C(T[] a)
    {
        var span = new Span<T>(a);
        T result = span[span.Length / 2];
        return result;
    }
}

class Y
{
    public static int Main(string[] args)
    {
        string[] vals = new string[] {"a", "b", "c", "d", "e"};
        string z = X<string>.C(vals);
        return z.Equals("c") ? 100 : -1;
    }
}
```
When `T` is a value type `X<T>` is not shared and we can inline the span methods. Eg for `X<int>`:
```
Inlines into 06000001 X`1[Int32][System.Int32]:C(ref):int
  [1 IL=0003 TR=000004 06000DB0] [aggressive inline attribute] System.Span`1[Int32][System.Int32]:.ctor(ref):this
    [0 IL=0004 TR=000086 060001F4] [FAILED: does not return] System.ThrowHelper:ThrowArgumentNullException(int)
    [2 IL=0055 TR=000057 06003F11] [below ALWAYS_INLINE size] System.Runtime.CompilerServices.JitHelpers:GetArrayData(ref):byref
  [3 IL=0012 TR=000011 06000DB7] [below ALWAYS_INLINE size] System.Span`1[Int32][System.Int32]:get_Length():int:this
  [4 IL=0019 TR=000016 06000DB9] [aggressive inline attribute] System.Span`1[Int32][System.Int32]:get_Item(int):byref:this
    [0 IL=0009 TR=000129 060001E4] [FAILED: does not return] System.ThrowHelper:ThrowIndexOutOfRangeException()
    [5 IL=0029 TR=000123 06003F15] [aggressive inline attribute] System.Runtime.CompilerServices.Unsafe:Add(byref,int):byref
```
But when `T` is a ref type `X<T>` is shared and the inlines are blocked, eg for `X<string>`
```
Inlines into 06000001 X`1[__Canon][System.__Canon]:C(ref):ref
  [0 IL=0003 TR=000004 06000DB0] [FAILED: runtime dictionary lookup] System.Span`1[__Canon][System.__Canon]:.ctor(ref):this
  [1 IL=0012 TR=000039 06000DB7] [below ALWAYS_INLINE size] System.Span`1[__Canon][System.__Canon]:get_Length():int:this
  [0 IL=0019 TR=000072 06000DB9] [FAILED: runtime dictionary lookup] System.Span`1[__Canon][System.__Canon]:get_Item(int):byref:this
```

**Comment by AndyAyersMS** [comment]
The indexer (`get_Item`) can be inlined into a shared caller if `Unsafe.Add` is recognized by the jit as an intrinsic. I have [an implementation of this](https://github.com/AndyAyersMS/coreclr/tree/SharedSpanIntrinsics) working.

Even so, many other span methods instantiated over ref types will still fail to inline, and enabling these would require more methods to be handled as intrinsics, so that jit can see through to the point where it can be sure that there are no required runtime lookups from inlinees.

Selectively picking methods to accelerate will give Span uneven performance characteristics that are hard to reason about. Accelerating Span methods over ref types comprehensively will be a lot of work and will make it much more difficult to make changes in the implementation, since these methods will only be fully expressed in the jit's IR.

The indexer might be a special case that is worth fixing. I'm less sure about any of the other methods.

@KrzysztofCwalina elsewhere you said `Span<ref type>` perf is not all that interesting. Do you still feel that way? 

@jkotas @stephentoub @russellhadley any thoughts on this?

**Comment by jkotas** [comment]
> The indexer (`get_Item`) can be inlined into a shared caller if `Unsafe.Add` is recognized by the jit as an intrinsic.

I think we should go even one level up and treat the indexer as intrinsic. It expect that it will be needed to enable bounds check elimination.

> `Span<ref type>` perf is not all that interesting

It is not interesting for the current usage in Pipelines and kestrel. It is likely to become interesting in future once there is more code written against Span.

**Comment by russellhadley** [comment]
@JosephTremoulet has been looking at the bounds check angle in #10050. (and I recently moved it to 2.1 because struct issues have been more pressing)

**Comment by AndyAyersMS** [comment]
After #10910, the indexer is now an intrinsic so is effectively inlined in all cases.

I'm going to close this out. When/if some other Span API warrants special handling we can open a new issue for it...
